File: backend/app/api/v1/transactions.py
"""Transaction endpoints. Every route is authenticated and scoped to the
caller's own transactions via app.services.transaction_service - the
frontend never supplies a user_id, only the resource ids it wants to act on."""

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.core.database import get_db
from app.models.transaction import TransactionType
from app.models.user import User
from app.repositories.transaction_repository import SortDirection, SortField, TransactionFilters
from app.schemas.transaction import QuickAddParseRequest, TransactionCreate, TransactionUpdate
from app.services import transaction_service

router = APIRouter(prefix="/transactions", tags=["transactions"])

_MAX_PAGE_SIZE = 200


@router.get("", response_model=None)
async def list_transactions(
    q: str | None = None,
    category_id: UUID | None = None,
    account_id: UUID | None = None,
    type: TransactionType | None = None,  # noqa: A002 - matches the public query param name
    date_from: datetime | None = None,
    date_to: datetime | None = None,
    amount_min: int | None = None,
    amount_max: int | None = None,
    is_recurring: bool | None = None,
    tags: list[str] = Query(default_factory=list),
    sort_by: SortField = "occurred_at",
    sort_dir: SortDirection = "desc",
    limit: int = Query(default=50, ge=1, le=_MAX_PAGE_SIZE),
    offset: int = Query(default=0, ge=0),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    filters = TransactionFilters(
        search=q,
        category_id=category_id,
        account_id=account_id,
        type=type,
        date_from=date_from,
        date_to=date_to,
        amount_min=amount_min,
        amount_max=amount_max,
        is_recurring=is_recurring,
        tags=tags,
        sort_by=sort_by,
        sort_dir=sort_dir,
        limit=limit,
        offset=offset,
    )
    transactions, total = await transaction_service.list_transactions(
        db, user_id=current_user.id, filters=filters
    )
    return {
        "data": transactions,
        "error": None,
        "meta": {"count": total, "limit": limit, "offset": offset},
    }


@router.post("", response_model=None, status_code=201)
async def create_transaction(
    body: TransactionCreate,
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    transaction = await transaction_service.create_transaction(
        db, user_id=current_user.id, data=body, idempotency_key=idempotency_key
    )
    await db.commit()
    return {"data": transaction, "error": None, "meta": None}


@router.post("/quick-add/parse", response_model=None)
async def quick_add_parse(
    body: QuickAddParseRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    result = await transaction_service.quick_add_parse(db, user_id=current_user.id, text=body.text)
    return {"data": result, "error": None, "meta": None}


@router.get("/{transaction_id}", response_model=None)
async def get_transaction(
    transaction_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    transaction = await transaction_service.get_transaction(
        db, user_id=current_user.id, transaction_id=transaction_id
    )
    return {"data": transaction, "error": None, "meta": None}


@router.put("/{transaction_id}", response_model=None)
async def update_transaction(
    transaction_id: UUID,
    body: TransactionUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    transaction = await transaction_service.update_transaction(
        db, user_id=current_user.id, transaction_id=transaction_id, data=body
    )
    await db.commit()
    return {"data": transaction, "error": None, "meta": None}


@router.delete("/{transaction_id}", response_model=None, status_code=200)
async def delete_transaction(
    transaction_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    await transaction_service.delete_transaction(
        db, user_id=current_user.id, transaction_id=transaction_id
    )
    await db.commit()
    return {"data": {"success": True}, "error": None, "meta": None}


@router.post("/{transaction_id}/duplicate", response_model=None, status_code=201)
async def duplicate_transaction(
    transaction_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    transaction = await transaction_service.duplicate_transaction(
        db, user_id=current_user.id, transaction_id=transaction_id
    )
    await db.commit()
    return {"data": transaction, "error": None, "meta": None}

File: frontend/lib/transactions.ts
import { apiRequest, apiRequestWithMeta } from "@/lib/api-client";

export const TRANSACTION_TYPES = ["income", "expense", "transfer"] as const;
export type TransactionType = (typeof TRANSACTION_TYPES)[number];

export interface Transaction {
  id: string;
  account_id: string;
  transfer_account_id: string | null;
  category_id: string | null;
  type: TransactionType;
  amount_minor: number;
  currency: string;
  merchant: string | null;
  description: string | null;
  notes: string | null;
  payment_method: string | null;
  tags: string[];
  occurred_at: string;
  is_recurring: boolean;
  recurring_transaction_id: string | null;
  /** Set only for a transaction imported by the automatic sync feature
   * (see lib/sync.ts) - null for every manual, recurring-generated, or
   * receipt-derived transaction. */
  linked_account_id: string | null;
  /** True only for a low-confidence synced transaction awaiting user
   * review - see backend app.models.transaction. Always false otherwise. */
  needs_review: boolean;
  created_at: string;
  updated_at: string;
}

export interface TransactionInput {
  account_id: string;
  type: TransactionType;
  amount_minor: number;
  transfer_account_id?: string | null;
  category_id?: string | null;
  merchant?: string | null;
  description?: string | null;
  notes?: string | null;
  payment_method?: string | null;
  tags?: string[];
  occurred_at?: string;
  is_recurring?: boolean;
}

export type TransactionUpdateInput = Partial<TransactionInput> & {
  clear_category?: boolean;
  /** Opt-in only - persists the transaction's (merchant, category) pair as
   * a reusable rule (see lib/merchant-rules.ts) applied to future synced
   * transactions from the same merchant. Never set unless the user
   * explicitly asked to remember it on this specific edit. */
  remember_category_for_merchant?: boolean;
};

export interface TransactionFilters {
  q?: string;
  category_id?: string;
  account_id?: string;
  type?: TransactionType;
  date_from?: string;
  date_to?: string;
  amount_min?: number;
  amount_max?: number;
  is_recurring?: boolean;
  tags?: string[];
  sort_by?: "occurred_at" | "amount_minor" | "created_at";
  sort_dir?: "asc" | "desc";
  limit?: number;
  offset?: number;
}

export interface TransactionListResult {
  transactions: Transaction[];
  total: number;
}

export interface QuickAddParseResult {
  raw_text: string;
  amount_minor: number | null;
  category_id: string | null;
  category_name: string | null;
  description: string | null;
  occurred_at: string;
  confidence: "high" | "medium" | "low";
  error: string | null;
}

function buildQuery(filters: TransactionFilters): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value === undefined || value === null || value === "") continue;
    if (Array.isArray(value)) {
      value.forEach((v) => params.append(key, String(v)));
    } else {
      params.set(key, String(value));
    }
  }
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function listTransactions(
  accessToken: string,
  filters: TransactionFilters = {}
): Promise<TransactionListResult> {
  const { data, meta } = await apiRequestWithMeta<Transaction[]>(
    `/api/v1/transactions${buildQuery(filters)}`,
    { accessToken }
  );
  return { transactions: data, total: (meta?.count as number) ?? data.length };
}

export async function createTransaction(
  accessToken: string,
  input: TransactionInput,
  idempotencyKey?: string
): Promise<Transaction> {
  return apiRequest<Transaction>("/api/v1/transactions", {
    method: "POST",
    accessToken,
    body: input,
    idempotencyKey,
  });
}

export async function updateTransaction(
  accessToken: string,
  transactionId: string,
  input: TransactionUpdateInput
): Promise<Transaction> {
  return apiRequest<Transaction>(`/api/v1/transactions/${transactionId}`, {
    method: "PUT",
    accessToken,
    body: input,
  });
}

export async function deleteTransaction(accessToken: string, transactionId: string): Promise<void> {
  await apiRequest(`/api/v1/transactions/${transactionId}`, { method: "DELETE", accessToken });
}

export async function duplicateTransaction(
  accessToken: string,
  transactionId: string
): Promise<Transaction> {
  return apiRequest<Transaction>(`/api/v1/transactions/${transactionId}/duplicate`, {
    method: "POST",
    accessToken,
  });
}

export async function parseQuickAdd(
  accessToken: string,
  text: string
): Promise<QuickAddParseResult> {
  return apiRequest<QuickAddParseResult>("/api/v1/transactions/quick-add/parse", {
    method: "POST",
    accessToken,
    body: { text },
  });
}
